test/base/src/androidTest/kotlin/io/realm/test/shared/EmbeddedObjectTests.kt

```diff
@@ -0,0 +1,1161 @@
+    //         }
+    //     }
+    // }
+    //
+
+    // NOT RELEVANT - Not possible through copyToRealm API
+    // @org.junit.Test
+    // fun dynamicRealm_createEmbeddedObject_wrongParentPropertyObjectTypeThrows() =
+    //     DynamicRealm.getInstance(realm.configuration).use { realm ->
+    //         realm.executeTransaction {
+    //             val parent = realm.createObject("EmbeddedSimpleParent", "parent")
+    //
+    //             assertFailsWith<IllegalArgumentException> {
+    //                 // Embedded object is not of the type the parent object links to.
+    //                 realm.createEmbeddedObject("EmbeddedTreeLeaf", parent, "child")
+    //             }
+    //         }
+    //     }
+    //
+
+    // NOT RELEVANT - Not possible through copyToRealm API
+    // @org.junit.Test
+    // fun dynamicRealm_createEmbeddedObject_wrongParentPropertyListTypeThrows() =
+    //     DynamicRealm.getInstance(realm.configuration).use { realm ->
+    //         realm.executeTransaction {
+    //             val parent = realm.createObject("EmbeddedSimpleListParent", "parent")
+    //
+    //             assertFailsWith<IllegalArgumentException> {
+    //                 // Embedded object is not of the type the parent object links to.
+    //                 realm.createEmbeddedObject("EmbeddedTreeLeaf", parent, "children")
+    //             }
+    //         }
+    //     }
+    //
+
+    // NOT RELEVANT - Not possible with EmbeddedObject
+    // @org.junit.Test
+    // fun copyToRealm_noParentThrows() = realm.executeTransaction {
+    //     assertFailsWith<IllegalArgumentException> {
+    //         realm.copyToRealm(EmbeddedSimpleChild("child"))
+    //     }
+    // }
+    //
+
+    // NOT RELEVANT - Not possible with EmbeddedObject
+    // @org.junit.Test
+    // fun copyToRealmOrUpdate_NoParentThrows() = realm.executeTransaction {
+    //     assertFailsWith<IllegalArgumentException> {
+    //         realm.copyToRealmOrUpdate(EmbeddedSimpleChild("child"))
+    //     }
+    // }
+    //
+
+    @Test
+    fun copyToRealm_child() {
+        realm.writeBlocking {
+            val parent = EmbeddedParent()
+            parent.child = EmbeddedChild()
+            copyToRealm(parent)
+        }
+
+        realm.query<EmbeddedParent>().find().single()
+        realm.query<EmbeddedChild>().find().single()
+    }
+
+    @Test
+    fun copyToRealm_childList() {
+        realm.writeBlocking {
+            val parent = EmbeddedParent()
+            val child = EmbeddedChild()
+            parent.child = child
+            parent.childList = realmListOf(child, child)
+            copyToRealm(parent)
+        }
+
+        realm.query<EmbeddedParent>().find().single()
+        realm.query<EmbeddedChild>().find().run {
+            // Every reference to an embedded object is cloned
+            assertEquals(3, size)
+        }
+    }
+
+    @Test
+    fun copyToRealm_treeSchema() {
```

Comment:
Would it make sense to also have tree schema with normal objects interleaved?

Reply:
I added a copyToRealm_tree_mixedRealmAndEmbeddedObject-test with a normal-embedded-normal-embedded chain of objects.